src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,267 @@
+        // Create a new layer for the operation.
+        Time layerStartTime = 0;
+        if (!this->metricsByLayer.empty())
+        {
+            const Layer& lastLayer = this->metricsByLayer.back();
+            layerStartTime = lastLayer.startTime + lastLayer.duration;
+        }
+        this->metricsByLayer.push_back(Layer{max(this->preferredLayerDuration, opDuration), layerStartTime});
+
+        return this->metricsByLayer.size() - 1;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::FindLayerToInsertOperationInto
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::FindLayerToInsertOperationInto(Qubit q, Duration opDuration) const
+    {
+        const QubitState& qstate = this->UseQubit(q);
+
+        LayerId layerToInsertInto = INVALID;
+        if (qstate.layer != INVALID)
+        {
+            const Layer& lastUsedIn = this->metricsByLayer[qstate.layer];
+            if (qstate.lastUsedTime + opDuration <= lastUsedIn.startTime + lastUsedIn.duration)
+            {
+                layerToInsertInto = qstate.layer;
+            }
+            else
+            {
+                for (LayerId candidate = qstate.layer + 1; candidate < this->metricsByLayer.size(); candidate++)
+                {
+                    if (opDuration <= this->metricsByLayer[candidate].duration)
+                    {
+                        layerToInsertInto = candidate;
+                        break;
+                    }
+                }
+            }
+        }
+        else if (opDuration <= this->preferredLayerDuration && !this->metricsByLayer.empty())
+        {
+            // the qubit hasn't been used in any of the layers yet -- add it to the first layer
+            layerToInsertInto = 0;
+        }
+
+        if (layerToInsertInto != INVALID && this->globalBarrier != INVALID)
+        {
+            if (this->globalBarrier + 1 == this->metricsByLayer.size())
+            {
+                layerToInsertInto = INVALID;
+            }
+            else
+            {
+                layerToInsertInto = std::max(layerToInsertInto, this->globalBarrier + 1);
+            }
+        }
+
+        return layerToInsertInto;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::AddOperationToLayer
+    //------------------------------------------------------------------------------------------------------------------
+    void CTracer::AddOperationToLayer(OpId id, LayerId layer)
+    {
+        assert(layer < this->metricsByLayer.size());
+        auto inserted = this->metricsByLayer[layer].operations.insert({id, 1});
+        if (!inserted.second)
+        {
+            assert(inserted.first->first == id);
+            inserted.first->second += 1;
+        }

```

Comment:
Can't this be done as `this->metricsByLayer[layer].operations[id] += 1;`?

Reply:
It can, because integer values in the map are default-initialized to zero :)